src/protocol/sort/multi_bit_permutation.rs

```diff
@@ -0,0 +1,207 @@
+/// by K. Chida, K. Hamada, D. Ikarashi, R. Kikuchi, N. Kiribuchi, and B. Pinkas
+/// <https://eprint.iacr.org/2019/695.pdf>.
+///
+/// Protocol to compute a secret sharing of a permutation, after sorting on multiple bits `num_multi_bits`.
+/// At a high level, the protocol works as follows:
+/// 1. Start with a vector of list of `L*n` secret shares `[[x1_1 ... x1_n], .. , [xL_1 ... xL_n]]` where each is a secret sharing of either zero or one.
+///    Here, L is the number of multi bits which are processed together (`num_multi_bits`) and n is the number of records
+/// 2. Equality Bit Checker : For j in 0 to 2 pow `num_multi_bits`
+///    i. Get binary representation of j (B1 .. BL)
+///    ii. For i in `num_multi_bits`
+///      a. Locally compute `mult_inputs` as (Bi * `xi_j` + (1-Bi)(1- `xi_j`))
+///   iii. Multiply all `mult_inputs` for this j
+/// 3. Compute accumulated sum: For j in 0 to 2 pow `num_multi_bits`
+///    i. For each record
+///       a. Calculate accumulated `prefix_sum` = s + `mult_output`
+/// 4. Compute the final output using sum of products executed in parallel for each record.
+#[allow(dead_code)]
+pub async fn multi_bit_permutation<'a, F: Field, S: SecretSharing<F>, C: Context<F, Share = S>>(
+    ctx: C,
+    input: &[Vec<S>],
+) -> Result<Vec<S>, Error> {
+    let num_multi_bits = input.len();
+    assert!(num_multi_bits > 0);
+    let num_records = input[0].len();
+    let num_possible_bit_values = 2 << (num_multi_bits - 1);
+
+    // Equality bit checker: this checks if each secret shared record is equal to any of numbers between 0 and num_possible_bit_values
+    let mut equality_check_futures = Vec::with_capacity(num_possible_bit_values);
+    for j in 0..num_possible_bit_values {
+        let ctx = ctx.clone();
+        equality_check_futures.push(async move { get_bit_equality_checkers(j, input, ctx).await });
+    }
+
+    let equality_checks = try_join_all(equality_check_futures).await?;
+
+    // Compute accumulated sum
+    let mut prefix_sum = Vec::with_capacity(num_possible_bit_values * num_records);
+    let mut cumulative_sum = S::ZERO;
+    for equality_check in &equality_checks {
+        for check in equality_check {
+            cumulative_sum += check;
+            prefix_sum.push(cumulative_sum.clone());
+        }
+    }
+
+    // Take sum of products of output of equality check and accumulated sum
+    let mut permutation_futures = Vec::with_capacity(num_records);
+    for rec in 0..num_records {
+        let ctx = ctx.clone();
+        let mut sop_inputs = Vec::with_capacity(num_possible_bit_values);
+        for idx in 0..num_possible_bit_values {
+            sop_inputs.push((
+                &equality_checks[idx][rec],
+                &prefix_sum[idx * num_records + rec],
+            ));
+        }
+        permutation_futures.push(async move {
+            ctx.sum_of_products(RecordId::from(rec), sop_inputs.as_slice())
```

Comment:
Right now we are:

1. Making a new copy for `N*M` cumulative sums.
2. Copying all of these AND also `N*M` elements from equality_checks into short vectors of pairs.
I think we can skip step 1 by just writing the cumulative sums directly into the vectors of pairs.

Here is a sketch (that prolly won't compile)

```
let mut permutation_inputs = Vec::with_capacity(num_records);
for (j, equality_check) in equality_checks.iter().enumerate() {
    for (i, check) in equality_check.iter().enumerate() {
        cumulative_sum += check;
        permutation_inputs.push((
            &equality_checks[j][i],
            cumulative_sum.clone(),
        ));
    }
}
let mut permutation_futures = Vec::with_capacity(num_records);
for (i, row) in permutation_inputs.iter().enumerate() {
    permutation_futures.push(async move {
        ctx.sum_of_products(RecordId::from(i), row.as_slice());
     });
}
try_join_all(permutation_futures).await
```

Reply:
We are not copying the elements but their locations only for cumulative_sum

Reply:
I see, I was wrong - you're not copying just taking a reference. Ignore this comment.